### PR TITLE
Support loadBalancerClass in service reconciliation

### LINKED ISSUE #7691: Service xxx is invalid: spec.loadBalancerClass: Invalid value: "null"
## Bug Report
**What did you do?**
Given the logstash config 
```yaml
apiVersion: logstash.k8s.elastic.co/v1alpha1
kind: Logstash
metadata:
  name: ${stack_name}
  namespace: ${namespace}
spec:
  count: 1
  elasticsearchRefs:
    - name: ${stack_name}
      clusterName: ${stack_name}
  version: ${version}
  pipelines:
    - pipeline.id: gelf
      config.string: |
        input {
          gelf {
            port => 12201
            add_field => { "input_source" => "gelf" }
          }
        }
        output {
          elasticsearch {
            hosts => [ "$${MAIN_ES_HOSTS}" ]
            user => "$${MAIN_ES_USER}"
            password => "$${MAIN_ES_PASSWORD}"
            ssl_certificate_authorities => "$${MAIN_ES_SSL_CERTIFICATE_AUTHORITY}"
            index => "gelf-%%{+YYYY.MM.dd}"
          }
        }
  services:
    - name: beats
      service:
        spec:
          type: NodePort
          ports:
            - port: 5044
              name: "filebeat"
              protocol: TCP
              targetPort: 5044
    - name: udp
      service:
        spec:
          type: LoadBalancer
          ports:
            - port: 12201
              name: "filebeat"
              protocol: UDP
              targetPort: 12201
```
The reconciliation was failing with 
```
{"log.level":"info","@timestamp":"2024-04-05T11:15:26.411Z","log.logger":"logstash-controller","message":"Updating resource","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","iteration":"20","namespace":"logging","logstash_name":"main","kind":"Service","namespace":"logging","name":"main-ls-udp"}
{"log.level":"info","@timestamp":"2024-04-05T11:15:26.437Z","log.logger":"logstash-es","message":"Ending reconciliation run","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","iteration":"2","namespace":"logging","logstash_name":"main","took":0.099525177}
{"log.level":"info","@timestamp":"2024-04-05T11:15:26.437Z","log.logger":"logstash-es","message":"Starting reconciliation run","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","iteration":"3","namespace":"logging","logstash_name":"main"}
{"log.level":"info","@timestamp":"2024-04-05T11:15:26.437Z","log.logger":"logstash-controller","message":"Ending reconciliation run","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","iteration":"20","namespace":"logging","logstash_name":"main","took":0.026232342}
{"log.level":"error","@timestamp":"2024-04-05T11:15:26.448Z","log.logger":"manager.eck-operator","message":"Reconciler error","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","controller":"logstash-controller","object":{"name":"main","namespace":"logging"},"namespace":"logging","name":"main","reconcileID":"fd0c79d7-32a7-4846-bada-3e8afd619e62","error":"Service \"main-ls-udp\" is invalid: spec.loadBalancerClass: Invalid value: \"null\": may not change once set","errorCauses":[{"error":"Service \"main-ls-udp\" is invalid: spec.loadBalancerClass: Invalid value: \"null\": may not change once set"}],"error.stack_trace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.17.2/pkg/internal/controller/controller.go:227"}
{"log.level":"info","@timestamp":"2024-04-05T11:15:26.523Z","log.logger":"logstash-es","message":"Ending reconciliation run","service.version":"2.12.1+a56215f7","service.type":"eck","ecs.version":"1.4.0","iteration":"3","namespace":"logging","logstash_name":"main","took":0.086298936}
```

preventing any further reconciliation (i.e. pipelines changes was not reflected in the secrets). Since the error was silent (pod not crashing etc), it took a while to get noticed. 

In order to fix the issue I had to explicitely define the `loadBalancerClass` field
```
  services:
    - name: udp
      service:
        spec:
          type: LoadBalancer
          loadBalancerClass: service.k8s.aws/nlb
          ports:
            - port: 12201
              name: "filebeat"
              protocol: UDP
              targetPort: 12201
```

AFAIK, this is not a required field, so the lack of it should not prevent the creation of the service. 

**What did you see instead? Under which circumstances?**

**Environment**

* ECK version:

	2.12.1

* Kubernetes information:
AWS
```
Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.10-eks-508b6b3", GitCommit:"e99f7c75641f738090d483d988dc4a70001e01cf", GitTreeState:"clean", BuildDate:"2024-01-29T20:59:05Z", GoVersion:"go1.20.13", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by pebrc**
Fixes  #7691

From the k8s docs: 

 > // loadBalancerClass is the class of the load balancer implementation this Service belongs to.
	// If specified, the value of this field must be a label-style identifier, with an optional prefix,
	// e.g. "internal-vip" or "example.com/internal-vip". Unprefixed names are reserved for end-users.
	// This field can only be set when the Service type is 'LoadBalancer'. If not set, the default load
	// balancer implementation is used, today this is typically done through the cloud provider integration,
	// but should apply for any default implementation. If set, it is assumed that a load balancer
	// implementation is watching for Services with a matching class. Any default load balancer
	// implementation (e.g. cloud providers) should ignore Services that set this field.
	// This field can only be set when creating or updating a Service to type 'LoadBalancer'.
	// Once set, it can not be changed. This field will be wiped when a service is updated to a non 'LoadBalancer' type.

AWS EKS uses this field on EKS clusters > 1.24 that have upgraded to the new [AWS LoadBalancer Controller](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html)

Behaviour this PR should implement: 
* default the `loadBalancerClass` from the server side values if no value is set in the ECK service spec
* recreate the service if the value set in the ECK service spec differes from the server side value
* allow resetting the value only if the `type` of the service is no longer `LoadBalancer`

### CODE REVIEW

File: pkg/controller/common/service_control_test.go [hunk 1]
```diff
@@ -246,6 +247,34 @@ func Test_needsDelete(t *testing.T) {
 			},
 			want: false,
 		},
+		{
```
Comment (barkbay) [inline pkg/controller/common/service_control_test.go:250]:
```suggestion
		{
			name: "No need to recreate if LoadBalancerClass has not changed",
			args: args{
				expected: corev1.Service{Spec: corev1.ServiceSpec{
					Type:              corev1.ServiceTypeLoadBalancer,
					LoadBalancerClass: ptr.To("my-customer/lb"),
				}},
				reconciled: corev1.Service{Spec: corev1.ServiceSpec{
					Type:              corev1.ServiceTypeLoadBalancer,
					LoadBalancerClass: ptr.To("my-customer/lb"),
				}},
			},
			want: false,
		},
		{
```

File: pkg/controller/common/service_control.go [hunk 2]
```diff
@@ -71,6 +71,11 @@ func needsRecreate(expected, reconciled *corev1.Service) bool {
 		return true
 	}
 
+	// LoadBalancerClass is immutable once set if target type is load balancer
+	if expected.Spec.Type == corev1.ServiceTypeLoadBalancer && expected.Spec.LoadBalancerClass != reconciled.Spec.LoadBalancerClass {
```
Comment (barkbay) [inline pkg/controller/common/service_control.go:75]: I think these are pointers, not values.

```suggestion
	if expected.Spec.Type == corev1.ServiceTypeLoadBalancer && !reflect.DeepEqual(expected.Spec.LoadBalancerClass, reconciled.Spec.LoadBalancerClass) {
```

**Review by barkbay — CHANGES_REQUESTED** [review]

**Review by barkbay — COMMENTED** [review]
I'm also wondering if we should not also update `applyServerSideValues` for `ExternalTrafficPolicy` and `AllocateLoadBalancerNodePorts`, I think `needsUpdate(...)` always returns `true` on my test cluster with the following diff:

```diff
  v1.ServiceSpec{
        ... // 8 identical fields
        LoadBalancerSourceRanges:      nil,
        ExternalName:                  "",
-       ExternalTrafficPolicy:         "",
+       ExternalTrafficPolicy:         "Cluster",
        HealthCheckNodePort:           0,
        PublishNotReadyAddresses:      false,
        SessionAffinityConfig:         nil,
        IPFamilies:                    {"IPv4"},
        IPFamilyPolicy:                &"SingleStack",
-       AllocateLoadBalancerNodePorts: nil,
+       AllocateLoadBalancerNodePorts: &true,
        LoadBalancerClass:             &"foo",
        InternalTrafficPolicy:         &"Cluster",
  }
```

File: pkg/controller/common/service_control.go (hunk 2 above)
Comment (pebrc) [reply pkg/controller/common/service_control.go:75]: 🤦 good catch

**Comment by pebrc** [comment]
> I'm also wondering if we should not also update applyServerSideValues for ExternalTrafficPolicy and AllocateLoadBalancerNodePorts

I think that's a good idea. I will add it to this PR

**Review by barkbay — APPROVED** [review]
LGTM!

**Pull request merged**